### ISSUE TITLE
[D8][UX] The admin role should have all permissions assigned to it

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1462]
I just created a new content type and then (by accident) happened to notice that the admin role did not have permission for any of the CRUD operations of that content type. Is there any reason for this? I cannot imagine any use case where the admin would be denied these permissions (especially since they are the ones creating the content types - ironic).

Actually, I don't see any reason why the site admins should be denied _any_ permission at all.

If there is a reason behind this behavior, then there is a bug: despite the permissions not assigned to the admin role, I could still create/edit/delete content of that type. UX WTF?.

So?

PS: related d.org issues:

[Make Admin Role Have All Permissions [#1860644]](https://www.drupal.org/node/1860644) (filed against the 7.x version of the [Panopoly](https://www.drupal.org/project/panopoly) distribution - fixed)

[New content types (and other dynamic) permissions are not auto-selected for admin role—always give it all perms [#1153072]](https://www.drupal.org/node/1153072) (filed against D8 core - closed as "works as designed" pointing to issue #2435075)

[Implement admin role as a flag on the role storage, simplify permissions page, remove user_modules_installed [#2435075]](https://www.drupal.org/node/2435075) (against D8 core - Change notice: https://www.drupal.org/node/2444095 )

[Dynamic permissions cannot be automatically assigned to or removed from roles [#787152]](https://www.drupal.org/node/787152) (filed against D7 core - fixed?)

...and a related module that does this: [Admin Role](https://www.drupal.org/project/adminrole)

> This module is a little helper to maintain an administrator role which has all available permissions.
> 
> [#D7CX](http://drupal.org/project/modules?solrsort=sort_title%20asc&text=d7cx): We don't need to pledge that Admin Role will have a full Drupal 7 release on the day that Drupal 7 is released because [Admin Role is part of Drupal 7](http://drupal.org/node/480660)! - **NOTE: as the core function that was added in D7 seems to fall short of expectations; this module has been ported to D7**.

![d7-admin_role](https://cloud.githubusercontent.com/assets/2423362/12024028/252f43a2-adf4-11e5-91a2-d3fb28dd991b.png)

> This role will be automatically assigned new permissions whenever a module is enabled, field is added/updated or entity is added/updated.

**Comment by klonos** [comment]
...just tested the same thing (creating a new content type) in D8 and the admin role automatically got all permissions for all the actions for that content type.

It seems that the way they've implemented it is that the admin role gets all permissions by default and all the checkboxes for all permissions in ```/admin/people/permissions``` are in a disabled state. So the admin role simply cannot be denied any permission. Makes me wonder why the admin role is listed in that page.

**Comment by klonos** [comment]
> ...Makes me wonder why the admin role is listed in that page.

Answering my own question here after I've read the related issue: without the admin role shown, the user has no visual clue that this role has all permissions assigned to it.

**Comment by Graham-72** [comment]
I believe I have seen in some places  that user #1 gets full permissions   because he/she is assumed to be the site administrator,  whereas  the administrator role can be assigned  to  more  than one person. This suggests that site administrator is a special case. But have I got this wrong?

**Comment by klonos** [comment]
I think you've got this right pretty much...

Administrator is a role and as such it can be assigned to many users. user 1 gets assigned the administrator role by default out of the box. Now, what I personally expect (and what obviously D7 user studies have shown most users expect) is that the Admin role is some kind of superuser (think linux root) that should automatically be assigned *all* permissions, current and future ones.

The situation with Drupal 7 and previous versions (as well as with Backdrop of course) is that only the permissions that are present after installation are assigned to this role. Any additional permissions added due to configuration (for example the use case of new content types created) or due to installation of new modules, are not auto-assigned to this role (unless of course the module developers have accounted for this and make it happen as part of the installation of their module).

More UX weirdness comes into play as users with the admin role are able to do tasks that they seemingly do not have permissions for.

**Comment by klonos** [comment]
This is part of #378

**Comment by klonos** [comment]
I think that https://www.drupal.org/node/787152 sums it up perfectly (That issue is about points 2 and 3 only):

> The admin role that comes configured in the standard profile has the following unexpected behaviors:

> 1.Permissions of certain (core) modules are not granted to the admin role when these modules are installed.
> 2. Permissions for new configuration items are not granted to the admin role.
> 3. Permissions introduced in a module update are not granted to the admin role.
> 4. Permissions of existing configuration items are not removed (from any role) when the configuration items are removed.

**Comment by Graham-72** [comment]
I find it a bit confusing and have to remind myself that the admin role is not the site administrator. I have never yet experienced user 1 not having full permissions.

**Comment by quicksketch** [comment]
Thanks @klonos for opening this issue. I've had this same experience with content types not getting new permissions.

This is definitely a bug. It was easy to overlook with content types (and taxonomy) in D7 because users that create content types have to have the "administer content" permission, so they also can create/edit/delete content as well. In Backdrop we separated "administer content" from "administer content types", so it's entirely possible for a non-UID-1 admin to create a new content type and then immediately not be able to create content of that type.

> I have never yet experienced user 1 not having full permissions. 

Yeah, UID 1 has a special check in `user_access()` to always allow access. So if you are UID 1, it's unlikely you'll have any problems. The entire point of the Admin Role however was to make it so that multiple users could share the responsibility of full permissions without being logged in as UID 1.

**Comment by klonos** [comment]
> This is definitely a bug....

Yeah, I was hesitant to adding that label, but with your permission...

> The entire point of the Admin Role however was to make it so that multiple users could share the responsibility of full permissions without being logged in as UID 1.

...and the entire point of the D8 implementation UI-wise is to visually denote that the admin role has *all* permissions (makes sure that *all* checkboxes in the admin column of the permissions page are *always* checked) and that this cannot change (hence the disabled/locked state of the checkboxes). This solves the UX WTF.

**Comment by Graham-72** [comment]
> Yeah, UID 1 has a special check in user_access() to always allow access. So if you are UID 1, it's unlikely you'll have any problems. The entire point of the Admin Role however was to make it so that multiple users could share the responsibility of full permissions without being logged in as UID 1.

Thanks for the clarification. There have been past instances where, in granting the Admin role, I have made use of the fact that the Admin Role has selected privileges rather than the full permission of UID 1. Is this now going to change for all permissions? If so, I'm thinking of consequential changes and also that I may have to rethink part of the Simple Access contrib module I recently ported.

**Comment by docwilmot** [comment]
> UID 1 has a special check in user_access() to always allow access

Perhaps we should do the same for admin roles. D8 seems to have done that; added a `isAdmin` flag to role configs and grant all access to those with that boolean set true. I think this should be the expected behavior for an admin role. If you want a `nearly-admin-role`, you can create your own and grant whatever limited permissions you want. 

So change:
```
function user_access($permission_name, $account = NULL) {
...
  // User #1 has all privileges:
  if ($account->uid == 1) {
    return TRUE;
  }
...
}
```

to:
```
function user_access($permission_name, $account = NULL) {
...
  $admin_role = config_get('system.core', 'user_admin_role');
  // User #1 and admin users have all privileges:
  if ($account->uid == 1 || in_array($admin_role, $account->roles)) {
    return TRUE;
  }
...
}

Then check and disable all admin permission checkboxes a la D8.

**Comment by klonos** [comment]
> ...If you want a nearly-admin-role, you can create your own and grant whatever limited permissions you want. 

@docwilmot :+1: to the whole thing. Reasoning and implementation.

**Comment by Dinornis** [comment]
@klonos I am unable to reproduce this in 1.5.2, perhaps this issue has been fixed already?

**Comment by klonos** [comment]
Closing this as obsolete (since the originally reported issue has in the meantime been addressed), and replacing it with #4372.
